docs: add NRI timeout and required_plugins guidance (#576)
rancher-sandbox/runtime-enforcer: docs/troubleshooting.adoc

If there is a discrepancy you will see a diff with the affected node and pods, followed by a full dump of the agent cache for that node.

== NRI timeouts and required plugins

Runtime Enforcer relies on NRI (Node Resource Interface) integration provided by the container runtime.

> Some environments may hit the NRI default request timeout (default is `2s`), which can cause the NRI plugin is detached from the container runtime. This can lead the agent to continuously retry to connect to the NRI plugin, resulting in the agent not coming up as expected.

Reviewer: ```suggestion
Some environments may hit the NRI default request timeout (default is `2s`), which can cause the NRI plugin to be detached from the container runtime. This can cause the agent to repeatedly retry connecting to the NRI plugin, resulting in the agent not coming up as expected.
```

> To mitigate this, you can increase NRI timeouts on the container runtime side by editing the container runtime configuration file.

Reviewer: ```suggestion
To mitigate this, you can increase the NRI timeout on the container runtime side by editing the container runtime configuration file.
```

  # plugin_request_timeout is the timeout for a plugin to handle an event/request.
  plugin_request_timeout = "2s"
----

For details and additional NRI settings, see https://github.com/containerd/containerd/blob/main/docs/NRI.md#disabling-nri-support-in-containerd[containerd NRI configuration documentation].

=== CRI-O: increase NRI timeout

Edit `/etc/crio/crio.conf` and update the `crio.nri` table:

[source,toml]
----
[crio.nri]
  # Timeout for a plugin to register itself with NRI.
  nri_plugin_registration_timeout = "5s"
  # Timeout for a plugin to handle an NRI request.
  nri_plugin_request_timeout = "2s"
----

For details, see https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.5.md#crionri-table[CRI-O NRI configuration documentation].

=== Required Plugins

Additionally, recent container runtimes provide an option to require a set of NRI plugins to always be present for container creation to succeed.

==== containerd: required_plugins

In `/etc/containerd/config.toml`, configure the `default_validator`:

[source,toml]
----
[plugins."io.containerd.nri.v1.nri".default_validator]
  enable = true
  required_plugins = [ <list of required NRI plugins> ]
----

> Replace `"<list of required NRI plugins>"` with the plugin name as registered with NRI in your environment.

Reviewer: Should we suggest the name of our plugin here? Not sure if we should put `00-runtime-enforcer-agent` or just  `runtime-enforcer-agent`. Moreover, we add a line saying that with this feature, we are sure that the pod cannot start without our agent running. The downside is that if the agent has some issues, no container will be started

Author: I think we can suggest out plugin name `runtime-enforcer-agent`, and add a warning to say that with this feature enabled, pod cannot start without our agent running.

See https://github.com/containerd/containerd/blob/main/docs/NRI.md#default-validator[containerd NRI configuration documentation] for more details.

==== CRI-O: nri_validator_required_plugins

In `/etc/crio/crio.conf`, configure the `crio.nri.default_validator` table:

[source,toml]
----
[crio.nri.default_validator]
  nri_enable_default_validator = true
  # List of required NRI plugins that must be present.
  nri_validator_required_plugins = ["<list of required NRI plugins>"]
----

See https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.5.md#crionridefault_validator-table[CRI-O NRI configuration documentation] for more details.

> NOTE: The *required plugins* feature is available since *containerd 2.2* (released Nov 5, 2025) and *CRI-O 1.34.0* (released Aug 27, 2025).

Reviewer 2: ```suggestion
NOTE: The *required plugins* feature is available since *containerd 2.2* and *CRI-O 1.34.0*.
```

I would drop these release dates 👍
